metrics/sinks/elasticsearch/ElasticsearchSink.cpp

```diff
@@ -0,0 +1,292 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2020 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "httpclient.hpp"
+#include "ElasticsearchSink.hpp"
+#include "IMeasurement.hpp"
+#include "IMetricSet.hpp"
+#include <algorithm>
+#include "jstring.hpp"
+
+
+using namespace hpccMetrics;
+
+extern "C" IMetricSink* getSinkInstance(const std::string& name, const IPropertyTree *pSettingsTree)
+{
+    IMetricSink *pSink = new ElasticsearchSink(name, pSettingsTree);
+    return pSink;
+}
+
+
+
+ElasticsearchSink::ElasticsearchSink(const std::string &name, const IPropertyTree *pSettingsTree) :
+    MetricSink(name, "elasticsearch")
+{
+    protocol.append("http");
+    port.append("9200");
+    domain.append("localhost");   // this will be a config parameter
+}
+
+
+void ElasticsearchSink::handle(const MeasurementVector &values, const std::shared_ptr<IMetricSet> &pMetricSet, MetricsReportContext *pContext)
+{
+
+    auto it = metricSets.find(pMetricSet->getName());
+    if (it != metricSets.end())
+    {
+        //
+        // Do this every time in case the index name changes because of something like a day change.
+        MetricSetInfo &curMetricSetInfo = metricSetInfo[pMetricSet->getName()];
+        initializeIndex(it->second);
+
+        //
+        // Build the JSON payload directly
+        std::string payload;
+        payload.append("{");
```

Comment:
This JSON structure seems simple enough, but you should consider utilizing JSON serialization tools for long term maintenance